Adds AI control beacons for exosuits (Winter Thaw 2017)

Xhuis opened the pull request:
_This feature was created as part of the [Winter Thaw of 2017](https://tgstation13.org/phpBB/viewtopic.php?f=10&t=9307) regardless of consensus or if it gets merged._

:cl: Xhuis
add: AI control beacons are a new item created from the exosuit fabricator. When installed into a mech, it allows AIs to jump to and from that mech freely. Note that malfunctioning AIs with the domination power unlocked will instead be forced to dominate the mech.
/:cl:

As requested by @GunHog.

* AIs can jump to and from the mech at will, unlike malfunctioning AIs that dominate mechs.
* Malfunctioning AIs must dominate mechs if applicable and cannot control one through the beacon.
* If the mech has a pilot, it cannot be controlled.
* Created from the exosuit fabricator with 1000 metal, 500 glass, and 200 silver. Research levels are programming 3, magnets 2, and engineering 2. (I think. I don't know how research code works, really.)

Commit:
Adds AI control beacons for exosuits

GunHog commented:
Oooooh I cannot wait to test this!!! :smiley: :smiley: :smiley:

GunHog commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -719,6 +719,14 @@
+	materials = list(MAT_METAL = 1000, MAT_GLASS = 500, MAT_SILVER = 200)
+	construction_time = 50
+	category = list("Misc")
+
```
You need to add `build_type = MECHFAB` in order to make it show up in the fabricator!

GunHog replied:
Additional note: You have not set the tech levels to the design! To do this, add this line:
`req_tech = list("programming" = 3, "magnets" = 2, "engineering" = 2)`

Xhuis replied:
I'll do this once I get home

GunHog replied:
May I make a PR to this branch? I am adding diagnostic HUD stuff as well.

GunHog commented:
Here are the Diag HUD sprites from @WJohn! Green is for normal beacon, blue is for the AI one!

![mech diag hud](https://cloud.githubusercontent.com/assets/4955510/22273188/aac5268a-e264-11e6-830b-d20b58fdfc33.png)

vcordie commented:
buffing silicons REEE

PKPenguin321 commented:
powerful creep

Xhuis commented:
Like I said, the feature doesn't matter to me. It got requested and was chosen by fair randomness, so I'm gonna make it anyway!

ShadowDeath6 commented:
can malf AIs return to their core after they've dominated a mech? if not then this would objectively be better than that which doesn't make much sense

Haevacht commented:
"Hey ai Can you take this ripley over to cargo? Law 2!" "I cannot" "AI MALF KILL IT"

But neat, AI piloted odys to prevent harm!

Mindustry commented:
So, if you build this, it will only be usable by non-antag AIs. What's to stop someone from building one every round if it's going to be a dead giveaway of a traitor?

XDTM commented:
Is the Malf AI not being abe to drive them a code-related issue or just a design choice?

GunHog commented:
I believe you (or Xhuis) misunderstand(s) the code. The `can_dominate_mechs` variable is only enabled once a Malf buys and activates the module. 

**A Malf AI can use an AI mech beacon for control as long as it has not activated the domination module.** It is not a dead giveaway unless the AI has enabled the domination module, which allows it to _steal_ mechs instead of taking ones that were specifically given to it. 

Again, to clarify how the code currently works: Any AI can use the mech beacon, including Traitor and upgraded AIs. The only exception is when an AI with malf powers _buys and activates the mech domination module._ This means that it uses domination mode instead of the beacon. It can still pilot the same mech, beacon or not. **AIs using the domination Malf Module cannot return to core. In exchange for being able to steal any mech, even piloted ones.** 

@Xhuis, if this is not your intent, then you need to adjust the code. In my opinion, the way it works now is perfectly fine!

Xhuis commented:
@ShadowDeath6 That's the case, but the AI is physically unable to control a mech unless a human installs a beacon in the exosuit.

@Mindustry There isn't.

@XDTM That's a choice. As long as they don't unlock the module, then they can control normally. The module lets them pilot a mech forcefully.

@GunHog You are correct, and that was my intent. About the text, I'm not sure how to fix it,as it was that way by default.

Commit:
Diagnostic HUD entry + Design Fix

- Adds Diagnostic HUD entries for mecha beacons. (Green for normal, Blue
for AI)
- Fixes AI beacon research designs to havep roper requirements

GunHog commented:
https://github.com/Xhuis/tgfork/pull/3 PR to your PR! I did not touch the beacon text, I am not sure what the best way of doing that would be, so I save it for later.

XDTM commented:
I misunderstood what you meant about Malf AIs then. I really like this, can't wait for it to be merged.

ShadowDeath6 commented:
So this is better than dominate mech in every way, considering you lose access to your core and malf AIs need to spend APC points on it.

GunHog commented:
That is true, and I can improve that later. I actually wanted the AI to have a return to core power, but deconstruction of a core was not possible when I made that feature. Also, I want to leave the AI's icon on the new core so people know that it belongs to an AI.

If this is merged, I will see about improving the Malf one.

RandomMarine commented:
Just don't let a malf AI be unable to do something that standard AIs can.

GunHog commented:
That is what I just said and intend to do, Marine. Again, to clarify: **A Malf AI with the domination power can still steal a mech loaded with an AI beacon. The only difference is that it cannot leave.** I will address this in another PR.

Being Malf does *not* prevent an AI using the beacon. Being Malf and having the Domination power simply means that you must steal mechs using that power instead of the beacon. You can still steal the same mech.

Commit:
Merge pull request #3 from GunHog/Xhuis_Slimy_Beacons

Diagnostic HUD entry + Design Fix

GunHog commented:
Do not forget to add href checks to cover things like the AI being depowered, being in a card or another mech, the mech being off camera, etc.

Xhuis commented:
It uses the same code for malf hacking, so if that has any checks then they'll carry ove.r

Commit:
Sanity checks for mech hijacking

GunHog commented:
Good guy slime cleaning up my mistakes! :stuck_out_tongue:
